### PR TITLE
Address code duplication in sha2-mb

### PR DESCRIPTION
**Pull request by msprotz**
## Proposed changes

Right now, sha2-mb duplicates a fair amount of code: update, update_blocks and update_last are identical for SHA-224/256 and SHA2-384/512 respectively.

This first PR demonstrates how to eliminate update_224 from Hacl.SHA2.Scalar32. I intend to do the same with update_nblocks_224 and update_last_224.

~As a "drive-by" change, it also removes more code in Hacl.Hash.SHA2 but this is orthogonal and something I want to do in a subsequent PR.~

A followup PR will take care of the 64-bit hashes -- right now, I'm focused on what's needed for the Python PR.

## Types of changes

What types of changes does your code introduce to HACL*?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)

## Notes

I have been careful to not change the interface of Hacl.Hash.SHA2 -- I will take care of all of that (or someone else will!) once I address #738

### CODE REVIEW

File: code/sha2-mb/Hacl.Spec.SHA2.EquivScalar.fsti [hunk 1]
```diff
@@ -9,14 +9,31 @@ open Hacl.Spec.SHA2
 
 #set-options "--z3rlimit 50 --fuel 0 --ifuel 0"
 
+val update_lemma: a:sha2_alg -> block:block_t a -> hash:words_state a ->
+  Lemma (update a block hash == Spec.Agile.Hash.update a hash block)
+
+val repeat_blocks_multi_extensionality:
```
Comment (R1kM) [inline code/sha2-mb/Hacl.Spec.SHA2.EquivScalar.fsti:15]: Move to lib.Sequence.Lemmas?

**Review by R1kM — APPROVED** [review]
Thanks! This looks overall good to me, and I'm all for avoiding code duplication.
I left a few minor comments, since this seems to change the C API (in Hacl_Streaming_SHA2), could you also add an entry to CHANGES.md?

File: dist/gcc-compatible/Hacl_Streaming_SHA2.c [hunk 2]
```diff
@@ -26,18 +26,18 @@
...
     (uint32_t)8U,
     (uint32_t)1U,
     uint32_t *os = hash;
-    uint32_t x = Hacl_Impl_SHA2_Generic_h224[i];
+    uint32_t x = Hacl_Impl_SHA2_Generic_h256[i];
     os[i] = x;);
 }
 
-static inline void sha224_update(uint8_t *b, uint32_t *hash)
+static inline void sha256_update0(uint8_t *b, uint32_t *hash)
```
[the remark this answers is not in the document]
Comment (msprotz) [reply dist/gcc-compatible/Hacl_Streaming_SHA2.c:40]: I think there's another sha256_update from Vale and krml doesn't have the whole-program analysis required to "know" that the two will never be in scope at the same time

File: dist/gcc-compatible/Hacl_Streaming_SHA2.c [hunk 3]
```diff
@@ -1076,10 +918,11 @@ void Hacl_Streaming_SHA2_free_256(Hacl_Streaming_SHA2_state_sha2_224 *s)
 /**
 Hash `input`, of len `input_len`, into `dst`, an array of 32 bytes.
 */
-void Hacl_Streaming_SHA2_sha256(uint8_t *dst, uint32_t input_len, uint8_t *input)
+void Hacl_Streaming_SHA2_sha256(uint8_t *input, uint32_t input_len, uint8_t *dst)
 {
+  uint8_t *dst1 = dst;
```
[the remark this answers is not in the document]
Comment (msprotz) [reply dist/gcc-compatible/Hacl_Streaming_SHA2.c:923]: fixed by my latest commit

File: code/sha2-mb/Hacl.Spec.SHA2.EquivScalar.fsti (hunk 1 above)
Comment (msprotz) [reply code/sha2-mb/Hacl.Spec.SHA2.EquivScalar.fsti:15]: done

File: dist/gcc-compatible/Hacl_Streaming_SHA2.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (msprotz) [reply dist/gcc-compatible/Hacl_Streaming_SHA2.c:40]: can you open up a followup issue if you think that's bothersome? I don't think this is something we can fix as part of this PR (unless we make sha224_update "the one")

**Comment by msprotz** [comment]
Drive-by change: fixup @R1kM's `find` invocation to exclude steel_types and now callconv.h (cf https://github.com/FStarLang/karamel/pull/320) while we're at it.

File: Makefile [hunk 4]
```diff
@@ -929,7 +929,7 @@ dist/test/c/Hacl_Test_K256.c: KRML_EXTRA=-drop Lib.IntTypes.Intrinsics -add-incl
 
 copy-krmllib:
 	mkdir -p dist/karamel
-	(cd $(KRML_HOME) && tar cvf - krmllib/dist/minimal $$(find include -not -name 'steel_types.h')) | (cd dist/karamel && tar xf -)
```
Comment (msprotz) [inline Makefile:932]: @R1kM the `find` invocation was also finding directories, so even though individual files were not included, the `include/krml` directory still appears, which tar interpreted as meaning "all the files in that directory")

File: code/sha2-mb/Hacl.SHA2.Scalar32.fst [hunk 5]
```diff
@@ -12,24 +12,178 @@ open Lib.MultiBuffer
...
+
+[@CInline]
+val sha224_update_nblocks: update_nblocks_vec_t' SHA2_224 M32
+let sha224_update_nblocks len b st =
+  let open Lib.Sequence in
+  let h0 = ST.get () in
+  sha256_update_nblocks len b st;
+  let h1 = ST.get () in
+  begin
+    let st0: lseq (element_t SHA2_256 M32) 8 = as_seq h0 st in
```
Comment (polubelova) [inline code/sha2-mb/Hacl.SHA2.Scalar32.fst:110]: Can we move the  `state_spec_v_extensionality`, `hacl_spec_update_224_256`, `hacl_spec_update_nblocks_224_256`, `hacl_spec_update_last_224_256` lemmas along with their `_scalar32` versions to `Hacl.SHA2.Scalar32.Lemmas`? Thanks!

Comment (R1kM) [reply code/sha2-mb/Hacl.SHA2.Scalar32.fst]: Good point, thanks. I moved all the proofs from this module to Hacl.SHA2.Scalar32.Lemmas

**Review by polubelova — APPROVED** [review]
Thanks!

**Pull request merged**